File: python/pii/spans.py
from __future__ import annotations

from dataclasses import dataclass
from typing import List, Sequence, Tuple

@dataclass
class PiiSpan:
    start: int
    endExclusive: int
    label: str

def assemble_spans(
    labels: Sequence[str],
    offsets: Sequence[Tuple[int, int]],
    attention_mask: Sequence[int],
) -> List[PiiSpan]:
    out: List[PiiSpan] = []
    open_label: str | None = None
    open_start = -1
    open_end = -1

    def close():
        nonlocal open_label, open_start, open_end
        if open_label is not None and open_start >= 0 and open_end > open_start:
            out.append(PiiSpan(open_start, open_end, open_label))
        open_label = None
        open_start = -1
        open_end = -1

    for i, tag in enumerate(labels):
        if i >= len(attention_mask) or attention_mask[i] == 0:
            continue
        s, e = offsets[i]
        if e <= s:
            continue
        if tag == "O":
            close()
            continue
        dash = tag.find("-")
        if dash < 0:
            close()
            continue
        prefix = tag[:dash]
        cls = tag[dash + 1 :]

        if prefix == "B":
            close()
            open_label, open_start, open_end = cls, s, e
        elif prefix == "I":
            if open_label == cls:
                open_end = e
            else:
                close()
                open_label, open_start, open_end = cls, s, e
        elif prefix == "E":
            if open_label == cls:
                open_end = e
                close()
            else:
                close()
                out.append(PiiSpan(s, e, cls))
        elif prefix == "S":
            close()
            out.append(PiiSpan(s, e, cls))
        else:
            close()

    close()
    return out

File: go/internal/pii/spans.go
package pii

type PiiSpan struct {
	Start        int    `json:"start"`
	EndExclusive int    `json:"endExclusive"`
	Label        string `json:"label"`
}

func AssembleSpans(labels []string, offsets [][2]int, attentionMask []int) []PiiSpan {
	var out []PiiSpan
	var openLabel string
	openStart := -1
	openEnd := -1
	hasOpen := false

	closeSpan := func() {
		if hasOpen && openStart >= 0 && openEnd > openStart {
			out = append(out, PiiSpan{Start: openStart, EndExclusive: openEnd, Label: openLabel})
		}
		hasOpen = false
		openLabel = ""
		openStart = -1
		openEnd = -1
	}

	for i, tg := range labels {
		if i >= len(attentionMask) || attentionMask[i] == 0 {
			continue
		}
		s := offsets[i][0]
		e := offsets[i][1]
		if e <= s {
			continue
		}
		if tg == "O" {
			closeSpan()
			continue
		}
		dash := -1
		for k := 0; k < len(tg); k++ {
			if tg[k] == '-' {
				dash = k
				break
			}
		}
		if dash < 0 {
			closeSpan()
			continue
		}
		prefix := tg[:dash]
		cls := tg[dash+1:]

		switch prefix {
		case "B":
			closeSpan()
			openLabel = cls
			openStart = s
			openEnd = e
			hasOpen = true
		case "I":
			if hasOpen && openLabel == cls {
				openEnd = e
			} else {
				closeSpan()
				openLabel = cls
				openStart = s
				openEnd = e
				hasOpen = true
			}
		case "E":
			if hasOpen && openLabel == cls {
				openEnd = e
				closeSpan()
			} else {
				closeSpan()
				out = append(out, PiiSpan{Start: s, EndExclusive: e, Label: cls})
			}
		case "S":
			closeSpan()
			out = append(out, PiiSpan{Start: s, EndExclusive: e, Label: cls})
		default:
			closeSpan()
		}
	}

	closeSpan()
	if out == nil {
		return []PiiSpan{}
	}
	return out
}
